dhall/src/Dhall/Pretty/Internal.hs

```diff
@@ -1121,6 +1123,94 @@ prettyCharacterSet characterSet expression =
 
         prettyText t = literal (Pretty.pretty (escapeText_ t))
 
+
+-- | Prepare 'Chunks' for multi-line formatting by interpolating problematic
+-- character sequences.
+--
+-- >>> multilineChunks (Chunks [] "\n \tx")
+-- Chunks [("\n",TextLit (Chunks [] " \t"))] "x"
+-- >>> multilineChunks (Chunks [] "\n\NUL\b\f\t")
+-- Chunks [("\n",TextLit (Chunks [] "\NUL\b\f"))] "\t"
+multilineChunks :: Chunks s a -> Chunks s a
+multilineChunks = interpolateControlCharacters . interpolateLastLineLeadingWhitespace
+
+-- | Interpolate leading whitespace on the last line
```

Comment:
Why only on the last line?  My understanding is that you would need to interpolate leading whitespace for all lines.  In other words, for a double-quoted string like this:

```dhall
" \n\t\n"
```

... you would have to emit:

```dhall
${" "}
${"\t"}
```

Reply:
We could do that but it doesn't seem to be necessary. In my (possibly incorrect) understanding of the dedentation rules, the whitespace on all but the last line is "protected" by the indentation level of the opening `''`. Only the leading whitespace on the very last line is ignored by the parser.


Aside: I wonder whether we should change the parser (and the standard?) to keep the whitespace on the last line in the case of

```dhall
''

  x''
```

and to only ignore it when the last is all whitespace, e.g.

```dhall
''

  ''
```

Reply:
@sjakobi: You're assuming that the string literal came from parsing a multi-line literal.  I'm talking about a situation where the user enters a double-quoted literal with newlines and each line is preceded by trailing whitespace.

Reply:
Oof, this stuff is confusing! :/

In any case, we can already handle your example fine on `master` – the whitespace is preserved:

```
$ echo '" \n\t\n"' | dhall
''
 
	
''
$ echo '" \n\t\n"' | dhall | dhall
''
 
	
''
Is this what you mean?

Reply:
@sjakobi: That particular example works because the leading whitespace is different, but now consider this case:

```dhall
" \n \n"
```

... or equivalently:

```dhall
''
${" "}
${" "}
''
```

Reply:
I don't get it – we're already handling this fine:

```
$ echo '" \\n \\n"' | dhall
''
 
 
''
$ echo '" \\n \\n"' | dhall | dhall
''
 
 
''
```

Reply:
@sjakobi: Oh, my mistake.  Now I see why you need to give the last line special treatment.  Ignore me, then! 🙂

Reply:
No worries @Gabriel439! I definitely don't understand the multi-line string spec 100%.